src/JustEat.StatsD/ConnectedSocketPool.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.CompilerServices;
+
+namespace JustEat.StatsD
+{
+    internal sealed class ConnectedSocketPool : IDisposable
+    {
+        private bool _disposed;
+        private readonly SimpleObjectPool<Socket> _pool;
+
+        public ConnectedSocketPool(IPEndPoint ipEndPoint)
+        {
+            IpEndPoint = ipEndPoint;
+
+            _pool = new SimpleObjectPool<Socket>(
+                Environment.ProcessorCount,
+                pool =>
+                {
+                    var socket = UdpTransport.CreateSocket();
+                    try
+                    {
+                        socket.Connect(ipEndPoint);
+                        return socket;
+                    }
+                    catch
+                    {
+                        socket?.Dispose();
+                        throw;
+                    }
+                });
+        }
+
+        public IPEndPoint IpEndPoint { get; }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Can you before & after the perf for with an without having to add these attributes? I'm curious as to what difference they make.

Reply:
Attribute did not make any difference because method was already inlined by compiler, but this is the difference between inlined

 |Method |     Mean |     Error |    StdDev | Allocated |
|------- |---------:|----------:|----------:|----------:|
|    Act | 79.10 ns | 0.7256 ns | 0.6433 ns |       0 B |

and MethodImplOptions.NoInlining methods

| Method |     Mean |    Error |   StdDev | Allocated |
|------- |---------:|---------:|---------:|----------:|
|    Act | 84.92 ns | 1.681 ns | 2.186 ns |       0 B |

Reply:
Code for benchmark was:

```csharp

    [MemoryDiagnoser]
    public class InliningBenchmark
    {
        private static readonly ConnectedSocketPool _csp =
            new ConnectedSocketPool(new IPEndPoint(IPAddress.Loopback, 1234));

        [Benchmark]
        public void Act()
        {
            var socket = _csp.PopOrCreate();
            _csp.Push(socket);
        }
    }

```